### ISSUE TITLE
amsthm, cleveref, \newtheorem, shared counter, \label

### ISSUE THREAD

**Issue by alexanderperlis** [latex3/tagging-project#773]
Thankfully `firstaid` seems to make `amsthm` and `cleveref` usable in many circumstances, allowing tagging experiments on existing documents.

Here's a particular combination that currently won't compile (on MacTeX, TeX Live 2024, both pdfTeX and LuaTeX, LaTeX2e 2024-11-01, L3 layer 2024-11-02, nor under `pdflatex-dev` or `lualatex-dev`, LaTeX2e 2025-06-01 pre-release-1 2024-12-18 branch, L3 layer 2024-12-09):
~~~~
\DocumentMetadata{testphase={phase-III,firstaid}}
\documentclass{article}
\usepackage{amsthm}
\usepackage{cleveref}

\newtheorem{theorem}{Theorem}[section]
\newtheorem{proposition}[theorem]{Proposition}

\begin{document}

\begin{proposition}
\label{mylabel}   % <--- error ("Something's wrong--perhaps a missing \item")
\end{proposition}

\end{document}
~~~~
Key ingredients:
- Loading `amsthm` before loading `cleveref` (as instructed in `cleverref` documentation)
- Loading `cleveref` before calling `\newtheorem` (as instructed in `cleverref` documentation)
- Defining a theorem type that uses someone else's counter
- Using `\label` with that theorem type

Error goes away with any one of the following (undesirable) changes:
- Comment-out `\DocumentMetadata` line
- Comment-out `\usepackage{amsthm}` line
- Comment-out `\usepackage{cleveref}` line
- Swap order of `\usepackage{amsthm}` and `\usepackage{cleveref}`
- Give `proposition` its own counter: `\newtheorem{proposition}{Proposition}[section]`
- Move `\usepackage{cleveref}` below the various `\newtheorem`

**Comment by krypt-n** [comment]
Quick notes (since I'm also interested in this combination of packages):

* the same compilation error happens for me without the `\label`

* using `lualatex-dev` (develop 2024-12-19 branch), I get a different error:

```
! LaTeX template Error: The instance '\l__block_block_instance_tl -\int_use:N
(template)              \g_block_nesting_depth_int ' of type 'block' is
(template)              unknown.
```

**Comment by u-fischer** [comment]
when cleveref is loaded after amsthm it overwrites the new tagging code for the theorems and so breaks tagging. Try what happens if you load it before amsthm, it is probably easier to solve whatever problem arisethere. 

Or consider to use e.g. zref-clever instead of cleveref. cleveref contains so many patches that is is quite hard to make it fully tagging compatible.

**Comment by alexanderperlis** [comment]
# Using `cleveref` with `amsthm`

Ulrike's suggestion to flip the order of `amsthm` and `cleveref`, also incorporating [Ulrike's naming patch given elsewhere](https://tex.stackexchange.com/a/730164/206306), seems to work in various current circumstances (with or without `\DocumentMetadata`, with any of `pdflatex`, `pdflatex-dev`, `lualatex`, `lualatex-dev`; LaTeX2e <2024-11-01> patch level 2 or LaTeX2e <2025-06-01> pre-release-1 (develop 2025-2-4 branch), and L3 programming layer <2025-01-18>).

~~~
\DocumentMetadata{testphase=latest}
\documentclass{article}
\usepackage{amsmath}
\usepackage{cleveref}  % load this before amsthm
\usepackage{amsthm}

\newtheorem{theorem}{Theorem}[section]
\newtheorem{proposition}[theorem]{Proposition}

% For next line, see https://tex.stackexchange.com/a/730164/206306
\AddToHook{env/proposition/begin}{\crefalias{theorem}{proposition}}

\crefname{theorem}{theorem}{theorems}
\crefname{proposition}{proposition}{propositions}

\begin{document}

\begin{proposition}
\label{mylabel}
\end{proposition}

Consider \cref{mylabel} and \Cref{mylabel}.

\end{document}
~~~

# Replacing `cleveref` with `zref-clever`

Ulrike's suggestion to instead use `zref-clever` struck me at first as relevant only when creating a *new* document; when attempting to add tagging to a large multi-file legacy document simply by adding `\DocumentMetadata` to the top, I presumed replacing `cleveref` would likely require an edit of every reference call throughout the various source files. This turned out not to be the case for the particular document I was working on. The author's preamble had `\usepackage[capitalise,noabbrev]{cleveref}`, along with calls to `\cref{somelabel}` throughout the various source files. For this particular type of simple `cleveref` usage, the switch to `zref-clever` was an easy change to the preamble:

~~~
#\usepackage[capitalise,noabbrev]{cleveref}
\usepackage{zref-clever}
\zcsetup{cap,noabbrev}
\def\cref#1{\zcref{#1}}
~~~

Actually, because the author had named their environments with abbreviated names, e.g., `theo` instead of `theorem`, I had to also add to the preamble a few additional lines to inform `zref-clever` of the singular/plural spellings of the various theorem-like environments:

~~~
\zcRefTypeSetup{theo}{
Name-sg = Theorem ,
name-sg = theorem ,
Name-pl = Theorems ,
name-pl = theorems ,
}
\zcRefTypeSetup{prop}{
Name-sg = Proposition ,
name-sg = proposition ,
Name-pl = Propositions ,
name-pl = propositions ,
}
~~~

In summary, the suggestion “consider using `zref-clever` instead of `cleveref`” turns out to be worthy of consideration not just for new documents but also for large multi-file legacy documents. It might turn out to be an easy replacement!
